Fix Docs for Modules and Refactor `@publicModule` tag to `@exportedAs`

petebacondarwin opened the pull request:
@naomiblack and @mhevery
- this PR fixes the module documentation issue. Now you can provide a jsdoc style comment block at the top of the module definition file that must include the `@module` tag to indicate that it documents the module rather than code components exported from the module.  Also you now specify that a module is or public consumption by applying the `@public` tag on the module itself. For example:

``` js
/**
 * @module
 * @public
 * @description
 * This block contains markdown that describe this module.
 */
```

The other major change in this PR is that the `@publicModule` tag has been replaced with the `@exportedAs` tag, which can be applied multiple times to a single code component. What this implies is that the associated code component has also been exported on other modules.

Closes #1258

petebacondarwin committed:
docs(test_lib/test_injector): fix invalid jsdoc type

petebacondarwin committed:
docs(*): add module description jsdoc tags

petebacondarwin committed:
chore(doc-gen): capture docs for modules from comments

Closes #1258

petebacondarwin committed:
docs(*): add @public tag to public modules

petebacondarwin committed:
chore(doc-gen): fix overview-dump template

The template was referencing an invalid property

mhevery commented on modules/angular2/change_detection.js:
```diff
@@ -45,7 +52,7 @@ export var defaultPipes = {
 
 /**
- * @publicModule angular2/change_detection
+ * @exportedFrom angular2/change_detection
```
Why the rename? I think `publicModule` is a better name. Can you share your thoughts?

mhevery replied:
I think it should be `exportedAs`

mhevery commented:
Rename `exportedFrom` to `exportedAs`

petebacondarwin commented:
The reason for the rename is that it no longer puts the onus on the exported component to tell that the module is public. This is also considerably more flexible as it allows modules that only contain direct exports and not reexports to be public. Also it allows items to be decl

petebacondarwin commented:
Phone finger slippage...

Also it allows components to be reexported internally too for the private docs

petebacondarwin commented:
Regarding the naming...

The tag is attached to a component that is an export on a private module. The tag declares that this component is also exported **from** another module. Exported **as** implies that a module is being aliased, which is not accurate.

But if you are sure you want it that way then I can change it.

mhevery commented:
I think `exportedAs` is still right. As it is being aliased.

Misko Hevery [10:03 AM]
it will be available from two places, it is just that we will not publish the original place.

petebacondarwin committed:
chore(doc-gen): use `@exportedAs` and `@public` rather than `@publicModule`

This commit refactors how we describe components that are re-exported in another
module. For example the "public" modules like `angular/angular` and `angular/annotations`
are public but they only re-export components from "private" modules.

Previously, you must apply the `@publicModule` tag to a component that was to be
re-exported. Applying this tag caused the destination module to become public.

Now, you specify that a module is public by applying the `@public` tag and then
you can "re-export" components to other modules by applying the `@exportedAs`
giving the name of the module from which the component will be re-exported.
tag. This tag can be used multiple times on a single component, allowing the
component to be exported on multiple modules.

petebacondarwin committed:
docs(*): rename `@publicModule` to `@exportedAs`

The `@publicModule` dgeni tag has been replaced by the `@exportedAs`
dgeni tag on components that are to be re-exported on another module.

petebacondarwin commented:
@mhevery - updated with `@exportedAs`. Feel free to get someone to merge once the rename PR has gone in.
